### PR TITLE
[V2] [WIP] fix: update cloud-provider-azure logic to handle preemption

### PR DESCRIPTION
**Pull request by sunpa93**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
* This PR is for testing the changes in `cloud-provider-azure` for handling `OperationPreempted` errors.

**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: vendor/sigs.k8s.io/cloud-provider-azure/pkg/provider/azure_controller_standard.go [hunk 1]
```diff
@@ -231,6 +231,27 @@ func (as *availabilitySet) UpdateVM(ctx context.Context, nodeName types.NodeName
...
+	klog.V(2).Infof("azureDisk - update(%s): vm(%s)", nodeResourceGroup, vmName)
+	// Invalidate the cache right after updating
+	defer func() {
+		_ = as.DeleteCacheForNode(vmName)
+	}()
+
+	future, rerr := as.VirtualMachinesClient.UpdateAsync(ctx, nodeResourceGroup, vmName, compute.VirtualMachineUpdate{}, "update_vm")
+	klog.V(2).Infof("azureDisk - update(%s): vm(%s) - returned with %v", nodeResourceGroup, vmName, rerr)
+	if rerr != nil {
+		return future, rerr.Error()
```
Comment (landreasyan) [inline vendor/sigs.k8s.io/cloud-provider-azure/pkg/provider/azure_controller_standard.go:250]: should we be returning nil instead of ` future` here?

Comment (landreasyan) [reply vendor/sigs.k8s.io/cloud-provider-azure/pkg/provider/azure_controller_standard.go:250]: This seems to be inconsistent from the way the rest of the code.

File: vendor/sigs.k8s.io/cloud-provider-azure/pkg/provider/azure_controller_vmssflex.go [hunk 2]
```diff
@@ -227,6 +227,28 @@ func (fs *FlexScaleSet) UpdateVM(ctx context.Context, nodeName types.NodeName) e
...
+	defer func() {
+		_ = fs.DeleteCacheForNode(vmName)
+	}()
+
+	klog.V(2).Infof("azureDisk - update(%s): vm(%s)", nodeResourceGroup, vmName)
+
+	future, rerr := fs.VirtualMachinesClient.UpdateAsync(ctx, nodeResourceGroup, vmName, compute.VirtualMachineUpdate{}, "update_vm")
+	klog.V(2).Infof("azureDisk - update(%s): vm(%s) - returned with %v", nodeResourceGroup, vmName, rerr)
+	if rerr != nil {
+		return future, rerr.Error()
```
Comment (landreasyan) [inline vendor/sigs.k8s.io/cloud-provider-azure/pkg/provider/azure_controller_vmssflex.go:247]: `return nil, rerr.Error()`?

File: vendor/sigs.k8s.io/cloud-provider-azure/pkg/provider/azure_controller_standard.go (hunk 1 above)
Comment (sunpa93) [reply vendor/sigs.k8s.io/cloud-provider-azure/pkg/provider/azure_controller_standard.go:250]: Although we were pretty sure that we wouldn't be getting an error with HTTP status code of 2xx for `UpdateAsync` calls, we just wanted to be safe and have a comprehensive way of capturing `PartialUpdateErrors` for these methods.

File: vendor/sigs.k8s.io/cloud-provider-azure/pkg/provider/azure_controller_vmss.go [hunk 3]
```diff
@@ -242,6 +242,34 @@ func (ss *ScaleSet) UpdateVM(ctx context.Context, nodeName types.NodeName) error
...
+	defer func() {
+		_ = ss.DeleteCacheForNode(vmName)
+	}()
+
+	klog.V(2).Infof("azureDisk - update(%s): vm(%s)", nodeResourceGroup, nodeName)
+	future, rerr := ss.VirtualMachineScaleSetVMsClient.UpdateAsync(ctx, nodeResourceGroup, vm.VMSSName, vm.InstanceID, compute.VirtualMachineScaleSetVM{}, "update_vmss_instance")
+
+	klog.V(2).Infof("azureDisk - update(%s): vm(%s) - returned with %v", nodeResourceGroup, nodeName, rerr)
+	if rerr != nil {
+		return future, rerr.Error()
```
[the remark this answers is not in the document]
Comment (sunpa93) [reply vendor/sigs.k8s.io/cloud-provider-azure/pkg/provider/azure_controller_vmss.go:268]: Please refer to the above reply on `azure_controller_`standard.go`

File: vendor/sigs.k8s.io/cloud-provider-azure/pkg/provider/azure_controller_common.go [hunk 4]
```diff
@@ -79,6 +88,7 @@ var defaultBackOff = kwait.Backoff{
 var (
 	managedDiskPathRE  = regexp.MustCompile(`.*/subscriptions/(?:.*)/resourceGroups/(?:.*)/providers/Microsoft.Compute/disks/(.+)`)
 	diskSnapshotPathRE = regexp.MustCompile(`.*/subscriptions/(?:.*)/resourceGroups/(?:.*)/providers/Microsoft.Compute/snapshots/(.+)`)
+	errorCodeRE        = regexp.MustCompile(`Code="(.*?)".*`)
```
Comment (landreasyan) [inline vendor/sigs.k8s.io/cloud-provider-azure/pkg/provider/azure_controller_common.go:91]: np: vmss refers to this as `resultCode`.

**Pull request closed without merging**